File: .agents/skills/pr-description-generator/scripts/generate_pr_description.py
#!/usr/bin/env python3

import argparse
import os
import re
import subprocess
import sys
from dataclasses import dataclass
from typing import List, Optional, Tuple


def _run_git(args: List[str], *, check: bool = True) -> subprocess.CompletedProcess:
    return subprocess.run(
        ["git", *args],
        check=check,
        text=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )


def _git_ok(args: List[str]) -> bool:
    try:
        _run_git(args, check=True)
        return True
    except subprocess.CalledProcessError:
        return False


def _detect_base_branch(preferred: str) -> str:
    candidates = [preferred]
    if preferred != "origin/develop":
        candidates.append("origin/develop")
    if preferred != "develop":
        candidates.append("develop")

    for c in candidates:
        if _git_ok(["rev-parse", "--verify", c]):
            return c

    raise RuntimeError(
        "Cannot find base branch. Tried: " + ", ".join(candidates) + ". "
        "Fetch remotes or pass --base explicitly."
    )


def _ensure_fetched(base: str, *, no_fetch: bool) -> None:
    if no_fetch:
        return
    if not base.startswith("origin/"):
        return
    # Best-effort fetch; keep it non-fatal to support offline use.
    try:
        _run_git(["fetch", "origin", "develop"], check=False)
    except Exception:
        pass


def _current_branch() -> str:
    cp = _run_git(["branch", "--show-current"], check=True)
    b = cp.stdout.strip()
    return b or "(detached)"


def _repo_root() -> str:
    cp = _run_git(["rev-parse", "--show-toplevel"], check=True)
    return cp.stdout.strip()


def _commit_subjects(base: str, head: str) -> List[str]:
    cp = _run_git(["log", "--no-merges", "--format=%s", f"{base}..{head}"])
    subjects = [s.strip() for s in cp.stdout.splitlines() if s.strip()]
    return subjects


def _changed_files(base: str, head: str) -> List[str]:
    cp = _run_git(["diff", "--name-status", f"{base}...{head}"])
    lines = [ln.rstrip() for ln in cp.stdout.splitlines() if ln.strip()]
    return lines


def _diffstat(base: str, head: str) -> str:
    cp = _run_git(["diff", "--stat", f"{base}...{head}"])
    return cp.stdout.rstrip()


def _is_migration_touched(changed_files: List[str]) -> bool:
    for ln in changed_files:
        parts = ln.split("\t")
        if not parts:
            continue
        path = parts[-1].strip()
        if path.startswith("drizzle/") and path.endswith(".sql"):
            return True
    return False


def _guess_title(subjects: List[str], branch: str) -> str:
    if subjects:
        return subjects[0]
    return f"PR: {branch}"


def _normalize_summary_bullet(s: str) -> str:
    s = s.strip()
    s = re.sub(r"\s+", " ", s)
    s = re.sub(r"\(#\d+\)$", "", s).strip()
    return s


def _summary_bullets(subjects: List[str]) -> List[str]:
    if not subjects:
        return ["No commits found between base and HEAD (check base branch)."]

    bullets: List[str] = []
    seen = set()
    for s in subjects:
        b = _normalize_summary_bullet(s)
        if not b:
            continue
        key = b.lower()
        if key in seen:
            continue
        seen.add(key)
        bullets.append(b)
        if len(bullets) >= 6:
            break

    return bullets


def _suggest_testing_commands() -> List[str]:
    cmds: List[str] = []
    if os.path.exists("mise.toml") or os.path.exists(".mise.toml"):
        cmds.append("mise run ci")
    if os.path.exists("pnpm-lock.yaml"):
        cmds.append("pnpm -s run ci:pr")
    if not cmds:
        cmds.append("<add your local test command>")
    return cmds


@dataclass
class PRData:
    title: str
    base: str
    head: str
    branch: str
    subjects: List[str]
    summary: List[str]
    changed_files: List[str]
    diffstat: str
    testing: List[str]
    notes: List[str]


def _build_pr_data(*, base: str, head: str, title: Optional[str], no_fetch: bool) -> PRData:
    _ensure_fetched(base, no_fetch=no_fetch)
    base = _detect_base_branch(base)

    branch = _current_branch()
    subjects = _commit_subjects(base, head)
    changed_files = _changed_files(base, head)
    diffstat = _diffstat(base, head)
    summary = _summary_bullets(subjects)

    notes: List[str] = []
    if _is_migration_touched(changed_files):
        notes.append("Includes DB migration changes; ensure migration applies cleanly.")

    final_title = title.strip() if title and title.strip() else _guess_title(subjects, branch)

    return PRData(
        title=final_title,
        base=base,
        head=head,
        branch=branch,
        subjects=subjects,
        summary=summary,
        changed_files=changed_files,
        diffstat=diffstat,
        testing=_suggest_testing_commands(),
        notes=notes,
    )


def _render_markdown(pr: PRData) -> str:
    lines: List[str] = []
    lines.append(f"# {pr.title}")
    lines.append("")
    lines.append("## Summary")
    for b in pr.summary:
        lines.append(f"- {b}")
    lines.append("")
    lines.append("## Changes")
    if pr.changed_files:
        lines.append("Changed files:")
        lines.append("")
        lines.append("```text")
        lines.extend(pr.changed_files)
        lines.append("```")
    else:
        lines.append("- No file changes detected (check base branch).")
    if pr.diffstat:
        lines.append("")
        lines.append("Diffstat:")
        lines.append("")
        lines.append("```text")
        lines.append(pr.diffstat)
        lines.append("```")

    lines.append("")
    lines.append("## Testing")
    for c in pr.testing:
        lines.append(f"- `{c}`")

    if pr.notes:
        lines.append("")
        lines.append("## Notes")
        for n in pr.notes:
            lines.append(f"- {n}")

    lines.append("")
    lines.append("---")
    lines.append("")
    lines.append(f"Base: `{pr.base}`")
    lines.append(f"Head: `{pr.head}`")
    lines.append(f"Branch: `{pr.branch}`")
    return "\n".join(lines).rstrip() + "\n"


def main() -> int:
    ap = argparse.ArgumentParser(
        description="Generate a copy-pastable PR description from git diff vs develop."
    )
    ap.add_argument(
        "--base",
        default="origin/develop",
        help="Base branch/ref to compare against (default: origin/develop).",
    )
    ap.add_argument(
        "--head",
        default="HEAD",
        help="Head ref to compare (default: HEAD).",
    )
    ap.add_argument("--title", default=None, help="Override PR title.")
    ap.add_argument(
        "--no-fetch",
        action="store_true",
        help="Do not fetch origin/develop before computing the diff.",
    )

    args = ap.parse_args()

    try:
        _repo_root()  # ensure we're in a git repo
        pr = _build_pr_data(
            base=args.base,
            head=args.head,
            title=args.title,
            no_fetch=args.no_fetch,
        )
    except Exception as e:
        sys.stderr.write(f"error: {e}\n")
        return 2

    md = _render_markdown(pr)
    sys.stdout.write("```markdown\n")
    sys.stdout.write(md)
    sys.stdout.write("```\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: .agents/skills/pr-description-generator/scripts/generate_pr_description.mjs
#!/usr/bin/env node

/*
  PR Description Generator

  Generates a copy-pastable markdown block containing:
  - Title (from first commit subject or override)
  - Summary bullets (deduped commit subjects)
  - Changes (name-status + diffstat)
  - Testing (repo-aware suggestions)
  - Notes (only when needed)
*/

import fs from "node:fs";
import { spawnSync } from "node:child_process";

function runGit(args, { check = true } = {}) {
  const cp = spawnSync("git", args, {
    encoding: "utf8",
    stdio: ["ignore", "pipe", "pipe"],
  });

  if (check && cp.status !== 0) {
    const err = new Error(cp.stderr?.trim() || `git ${args.join(" ")} failed`);
    err.exitCode = cp.status ?? 1;
    throw err;
  }

  return cp;
}

function gitOk(args) {
  try {
    runGit(args, { check: true });
    return true;
  } catch {
    return false;
  }
}

function detectBaseBranch(preferred) {
  const candidates = [preferred];
  if (preferred !== "origin/develop") candidates.push("origin/develop");
  if (preferred !== "develop") candidates.push("develop");

  for (const c of candidates) {
    if (gitOk(["rev-parse", "--verify", c])) return c;
  }

  throw new Error(`Cannot find base branch. Tried: ${candidates.join(", ")}. Fetch remotes or pass --base explicitly.`);
}

function ensureFetched(base, { noFetch }) {
  if (noFetch) return;
  if (!base.startsWith("origin/")) return;

  // Best-effort fetch; non-fatal for offline use.
  try {
    runGit(["fetch", "origin", "develop"], { check: false });
  } catch {
    // ignore
  }
}

function currentBranch() {
  const cp = runGit(["branch", "--show-current"]);
  const b = (cp.stdout || "").trim();
  return b || "(detached)";
}

function repoRoot() {
  const cp = runGit(["rev-parse", "--show-toplevel"]);
  return (cp.stdout || "").trim();
}

function commitSubjects(base, head) {
  const cp = runGit(["log", "--no-merges", "--format=%s", `${base}..${head}`]);
  return (cp.stdout || "")
    .split(/\r?\n/)
    .map((s) => s.trim())
    .filter(Boolean);
}

function changedFiles(base, head) {
  const cp = runGit(["diff", "--name-status", `${base}...${head}`]);
  return (cp.stdout || "")
    .split(/\r?\n/)
    .map((s) => s.replace(/\s+$/, ""))
    .filter((s) => s.trim());
}

function diffstat(base, head) {
  const cp = runGit(["diff", "--stat", `${base}...${head}`]);
  return (cp.stdout || "").replace(/\s+$/, "");
}

function isMigrationTouched(nameStatusLines) {
  for (const ln of nameStatusLines) {
    const parts = ln.split("\t");
    const p = (parts.at(-1) || "").trim();
    if (p.startsWith("drizzle/") && p.endsWith(".sql")) return true;
  }
  return false;
}

function guessTitle(subjects, branch) {
  if (subjects.length > 0) return subjects[0];
  return `PR: ${branch}`;
}

function normalizeSummaryBullet(s) {
  return s
    .trim()
    .replace(/\s+/g, " ")
    .replace(/\(#\d+\)$/, "")
    .trim();
}

function summaryBullets(subjects) {
  if (subjects.length === 0) {
    return ["No commits found between base and HEAD (check base branch)."];
  }

  const bullets = [];
  const seen = new Set();
  for (const s of subjects) {
    const b = normalizeSummaryBullet(s);
    if (!b) continue;
    const key = b.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    bullets.push(b);
    if (bullets.length >= 6) break;
  }
  return bullets;
}

function suggestTestingCommands() {
  const cmds = [];
  if (fs.existsSync("mise.toml") || fs.existsSync(".mise.toml")) cmds.push("mise run ci");
  if (fs.existsSync("pnpm-lock.yaml")) cmds.push("pnpm -s run ci:pr");
  if (cmds.length === 0) cmds.push("<add your local test command>");
  return cmds;
}

function renderMarkdown(pr) {
  const lines = [];
  lines.push(`# ${pr.title}`);
  lines.push("");
  lines.push("## Summary");
  for (const b of pr.summary) lines.push(`- ${b}`);

  lines.push("");
  lines.push("## Changes");
  if (pr.changedFiles.length > 0) {
    lines.push("Changed files:");
    lines.push("");
    lines.push("```text");
    lines.push(...pr.changedFiles);
    lines.push("```");
  } else {
    lines.push("- No file changes detected (check base branch).");
  }

  if (pr.diffstat) {
    lines.push("");
    lines.push("Diffstat:");
    lines.push("");
    lines.push("```text");
    lines.push(pr.diffstat);
    lines.push("```");
  }

  lines.push("");
  lines.push("## Testing");
  for (const c of pr.testing) lines.push(`- \`${c}\``);

  if (pr.notes.length > 0) {
    lines.push("");
    lines.push("## Notes");
    for (const n of pr.notes) lines.push(`- ${n}`);
  }

  lines.push("");
  lines.push("---");
  lines.push("");
  lines.push(`Base: \`${pr.base}\``);
  lines.push(`Head: \`${pr.head}\``);
  lines.push(`Branch: \`${pr.branch}\``);
  return `${lines.join("\n").replace(/\s+$/, "")}\n`;
}

function parseArgs(argv) {
  const out = {
    base: "origin/develop",
    head: "HEAD",
    title: null,
    noFetch: false,
  };

  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--base") out.base = argv[++i] || out.base;
    else if (a === "--head") out.head = argv[++i] || out.head;
    else if (a === "--title") out.title = argv[++i] || out.title;
    else if (a === "--no-fetch") out.noFetch = true;
    else if (a === "--help" || a === "-h") {
      out.help = true;
    } else {
      out.unknown = a;
    }
  }

  return out;
}

function printHelp() {
  process.stdout.write(`Usage: generate_pr_description.js [options]\n\n`);
  process.stdout.write(`Options:\n`);
  process.stdout.write(`  --base <ref>     Base ref to compare (default: origin/develop)\n`);
  process.stdout.write(`  --head <ref>     Head ref to compare (default: HEAD)\n`);
  process.stdout.write(`  --title <title>  Override PR title\n`);
  process.stdout.write(`  --no-fetch       Do not fetch origin/develop\n`);
  process.stdout.write(`  -h, --help       Show help\n`);
}

function main() {
  const args = parseArgs(process.argv);
  if (args.help) {
    printHelp();
    return 0;
  }
  if (args.unknown) {
    process.stderr.write(`error: unknown argument ${args.unknown}\n`);
    printHelp();
    return 2;
  }

  try {
    repoRoot(); // ensure we're in a git repo

    ensureFetched(args.base, { noFetch: args.noFetch });
    const base = detectBaseBranch(args.base);

    const branch = currentBranch();
    const subjects = commitSubjects(base, args.head);
    const changed = changedFiles(base, args.head);
    const stat = diffstat(base, args.head);
    const summary = summaryBullets(subjects);

    const notes = [];
    if (isMigrationTouched(changed)) {
      notes.push("Includes DB migration changes; ensure migration applies cleanly.");
    }

    const title = args.title && args.title.trim() ? args.title.trim() : guessTitle(subjects, branch);

    const md = renderMarkdown({
      title,
      base,
      head: args.head,
      branch,
      subjects,
      summary,
      changedFiles: changed,
      diffstat: stat,
      testing: suggestTestingCommands(),
      notes,
    });

    process.stdout.write("```markdown\n");
    process.stdout.write(md);
    process.stdout.write("```\n");
    return 0;
  } catch (e) {
    process.stderr.write(`error: ${e && e.message ? e.message : String(e)}\n`);
    return 2;
  }
}

process.exitCode = main();
